### PR TITLE
net-libs/libpsl: Added static-libs flag

### PR DESCRIPTION
**Pull request by lima-limon-inc**
---

Please check all the boxes that apply:

- [X] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [X] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [X] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [X] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

**Comment by lima-limon-inc** [comment]
Just want to clarify, the good'ol "works in my machine".

```
/usr
/usr/bin
/usr/bin/psl
/usr/bin/psl-make-dafsa
/usr/include
/usr/include/libpsl.h
/usr/lib
/usr/lib/debug
/usr/lib/debug/usr
/usr/lib/debug/usr/bin
/usr/lib/debug/usr/bin/psl.debug
/usr/lib/debug/usr/lib
/usr/lib/debug/usr/lib/libpsl.so.5.3.5.debug
/usr/lib/debug/usr/lib64
/usr/lib/debug/usr/lib64/libpsl.so.5.3.5.debug
/usr/lib/libpsl.a                          <--------- Here it is :wave: 
/usr/lib/libpsl.so -> libpsl.so.5
/usr/lib/libpsl.so.5 -> libpsl.so.5.3.5
/usr/lib/libpsl.so.5.3.5
/usr/lib/pkgconfig
/usr/lib/pkgconfig/libpsl.pc
/usr/lib64
/usr/lib64/libpsl.a
/usr/lib64/libpsl.so -> libpsl.so.5
/usr/lib64/libpsl.so.5 -> libpsl.so.5.3.5
/usr/lib64/libpsl.so.5.3.5
/usr/lib64/pkgconfig
/usr/lib64/pkgconfig/libpsl.pc
/usr/share
/usr/share/doc
/usr/share/doc/libpsl-0.21.5-r1
/usr/share/doc/libpsl-0.21.5-r1/AUTHORS.bz2
/usr/share/doc/libpsl-0.21.5-r1/ChangeLog.bz2
/usr/share/doc/libpsl-0.21.5-r1/NEWS.bz2
/usr/share/man
/usr/share/man/man1
/usr/share/man/man1/psl-make-dafsa.1.bz2
/usr/share/man/man1/psl.1.bz2
```

**Comment by lima-limon-inc** [comment]
Before the patch:

```
/usr
/usr/bin
/usr/bin/psl
/usr/bin/psl-make-dafsa
/usr/include
/usr/include/libpsl.h
/usr/lib
/usr/lib/debug
/usr/lib/debug/usr
/usr/lib/debug/usr/bin
/usr/lib/debug/usr/bin/psl.debug
/usr/lib/debug/usr/lib
/usr/lib/debug/usr/lib/libpsl.so.5.3.5.debug
/usr/lib/debug/usr/lib64
/usr/lib/debug/usr/lib64/libpsl.so.5.3.5.debug
/usr/lib/libpsl.so -> libpsl.so.5
/usr/lib/libpsl.so.5 -> libpsl.so.5.3.5
/usr/lib/libpsl.so.5.3.5
/usr/lib/pkgconfig
/usr/lib/pkgconfig/libpsl.pc
/usr/lib64
/usr/lib64/libpsl.so -> libpsl.so.5
/usr/lib64/libpsl.so.5 -> libpsl.so.5.3.5
/usr/lib64/libpsl.so.5.3.5
/usr/lib64/pkgconfig
/usr/lib64/pkgconfig/libpsl.pc
/usr/share
/usr/share/doc
/usr/share/doc/libpsl-0.21.5
/usr/share/doc/libpsl-0.21.5/AUTHORS.bz2
/usr/share/doc/libpsl-0.21.5/ChangeLog.bz2
/usr/share/doc/libpsl-0.21.5/NEWS.bz2
/usr/share/man
/usr/share/man/man1
/usr/share/man/man1/psl-make-dafsa.1.bz2
/usr/share/man/man1/psl.1.bz2
```

**Comment by eli-schwartz** [comment]
Procedural note on:
> Just want to clarify, the good'ol "works on my machine".

I would personally suggest using app-portage/iwdevtools with its /etc/portage/bashrc hooks including `qa-cmp`, that directly compare the before-and-after of a package and offer you a report that looks like this example for another package:

```
>>> Completed installing mail-client/thunderbird-bin-128.3.0 into /var/tmp/portage/mail-client/thunderbird-bin-128.3.0/image

 * Final size of build directory:      0 KiB
 * Final size of installed tree:  263736 KiB (257.5 MiB)

 * CMP: =mail-client/thunderbird-bin-115.15.0-r1 with mail-client/thunderbird-bin-128.3.0/image
 *  FILES:-opt/thunderbird/Throbber-small.gif
 *  FILES:-opt/thunderbird/crashreporter.ini
 *  FILES:+opt/thunderbird/libgkcodecs.so
 *  FILES:-opt/thunderbird/plugin-container
 *  SONAME:+libgkcodecs.so(64)
 * ------> FILES(+1,-3) SONAME(+1)
 * Updating .desktop files database ...                                                                                                                                                                     [ ok ]
 * Updating icons cache ...                                                                                                                                                                                 [ ok ]
 * Updating .desktop files database ...                                                                                                                                                                     [ ok ]
 * Updating icons cache ...                                                                                                                                                                                 [ ok ]
 * Optional programs for extra features:
 *   net-libs/libotr for encrypted chat support
 * VDB: detected possibly incorrect RDEPEND (mail-client/thunderbird-bin-128.3.0)
 * media-libs/libpulse < 
```

(Note this supports colors too! but a text copy/paste doesn't support them. :smile:)

It's significantly easier to read than "here is the filelist after, and here is the filelist before, so you can compare them." iwdevtools is *extremely* convenient. :smile:

**Comment by eli-schwartz** [comment]
On the topic of this actual change... your commit message describes *what* you are doing but not *why* you are doing it. So I think I should probably ask: why, in fact, are you doing it?

Many packages exist, most of which don't have USE=static-libs. Why this package specifically? Is there another package that needs it for something? e.g. a package that has USE=static and has to link statically to libpsl?

**Comment by lima-limon-inc** [comment]
> It's significantly easier to read than "here is the filelist after, and here is the filelist before, so you can compare them." iwdevtools is _extremely_ convenient. 😄

Oh! That's great! I was not aware of that tool. That definitely looks cleaner than pasting the "raw files" haha. Thank you very much :blush: 


> On the topic of this actual change... your commit message describes _what_ you are doing but not _why_ you are doing it. So I think I should probably ask: why, in fact, are you doing it?
> 
> Many packages exist, most of which don't have USE=static-libs. Why this package specifically? Is there another package that needs it for something? e.g. a package that has USE=static and has to link statically to libpsl?

Great point, I forgot to mention it. Here's the full thought process, I'll leave a TLDR at the end. 

I wanted to write an application that used `libcurl`. And, additionally, I wanted the application to be a single, statically linked executable.

This, as I soon figured out, turned out to be a **way** harder task than I had imagined. I turned to `pkg-config` for some guidance (plus some googling) into the libraries that I needed. And I got the following:

```shell
> pkg-config --cflags --static --libs libcurl
 -lcurl -lssl -lcrypto -lz -lssl -lcrypto -ldl -pthread -Wl,--push-state,--as-needed,-latomic,--pop-state -lpsl -lunistring -lidn2 -lunistring -lnghttp2 -lnghttp3 -lcares -lpthread
```

After fiddling around with `gcc` and it's flag ordering, I realized that I did not have all the `.a` libraries present in my system, so I decided to enable the `static-libs` USE flag in all the missing packages (`net-misc/curl`, `sys-libs/zlib`, `dev-libs/libunistring`, `net-dns/c-ares`, `net-dns/libidn2`, `net-libs/nghttp2`, `net-libs/nghttp3`, `net-libs/libpsl`).
The only library that was missing on Gentoo was `libpsl`. 

So after reading some of the Meson documentation (Which was pretty helpful actually! They have [a very useful chart](https://mesonbuild.com/Builtin-options.html) that pointed to towards the parameter I needed to pass in order to get both the static and dynamic library.

**TLDR:** I wanted to build a static executable that had libcurl statically. IIANM, `libpsl` was both needed and did not have a `static-libs` use flag.

There! I hope I made myself clear. 

Best regards! :blush:

**Comment by lima-limon-inc** [comment]
PS: If anyone's curious, in the end, I "managed" to compile the executable, although I did get the following warnings:

```
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64/libcrypto.a(libcrypto-lib-dso_dlfcn.o): in function `dlfcn_globallookup':
dso_dlfcn.c:(.text+0x15): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64/libcrypto.a(libcrypto-lib-bio_addr.o): in function `BIO_lookup_ex':
bio_addr.c:(.text+0xc2d): warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64/libcrypto.a(libcrypto-lib-bio_sock.o): in function `BIO_gethostbyname':
bio_sock.c:(.text+0x75): warning: Using 'gethostbyname' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64/libcurl.a(libcurl_la-netrc.o): in function `Curl_parsenetrc':
netrc.c:(.text+0x7fc): warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64/libcares.a(libcares_la-ares_getaddrinfo.o): in function `ares_getaddrinfo':
ares_getaddrinfo.c:(.text+0x8f3): warning: Using 'getservbyname' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../lib64/libcares.a(libcares_la-ares_getnameinfo.o): in function `lookup_service.part.0.constprop.0':
ares_getnameinfo.c:(.text+0x2dc): warning: Using 'getservbyport_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
``` 

I don't know if I messed up, or it's simply glib's nature/shortcomings. Time to look towards `muslc`?

**Comment by eli-schwartz** [comment]
> After fiddling around with `gcc` and it's flag ordering, I realized that I did not have all the `.a` libraries present in my system, so I decided to enable the `static-libs` USE flag in all the missing packages (`net-misc/curl`, `sys-libs/zlib`, `dev-libs/libunistring`, `net-dns/c-ares`, `net-dns/libidn2`, `net-libs/nghttp2`, `net-libs/nghttp3`, `net-libs/libpsl`).
> The only library that was missing on Gentoo was `libpsl`.

The fact that net-misc/curl has this USE flag, depends on libpsl, and libpsl lacks the same USE flag, feels like a compelling argument. Thanks for the explanation!

> So after reading some of the Meson documentation (Which was pretty helpful actually! They have [a very useful chart](https://mesonbuild.com/Builtin-options.html) that pointed to towards the parameter I needed to pass in order to get both the static and dynamic library.

As I am also a meson developer, thanks. :)



> I don't know if I messed up, or it's simply glib's nature/shortcomings. Time to look towards `muslc`?

This is basically weird because glibc wants to dlopen NSS libraries in order to support the use case where you compile an executable on your laptop and then scp it to a server running LDAP where it needs to query LDAP in order to figure out what user accounts are available so it can retrieve networked records for `getpwuid_r(32791)`.

You could also have an alternative DNS resolver that may be needed in practice to actually *resolve* stuff.

Anyway, there is some interesting conversation at https://developers.redhat.com/articles/2023/08/31/how-we-ensure-statically-linked-applications-stay-way

musl essentially "solves" this by not supporting those use cases, which may be a good thing in the event that it is mission-critical for you to avoid loading nss modules and it is acceptable to simply not support operation on systems where ignoring NSS leads to failure.

glibc defines a configure option:

> --enable-static-nss
> > Compile static versions of the NSS (Name Service Switch) libraries. This is not recommended because it defeats the purpose of NSS; a program linked statically with the NSS libraries cannot be dynamically reconfigured to use a different name database.

and the associated FAQ: https://sourceware.org/glibc/wiki/FAQ#Even_statically_linked_programs_need_some_shared_libraries_which_is_not_acceptable_for_me.__What_can_I_do.3F

**Comment by lima-limon-inc** [comment]
> The fact that net-misc/curl has this USE flag, depends on libpsl, and libpsl lacks the same USE flag, feels like a compelling argument. Thanks for the explanation!

On the contrary! :blush: 

> > So after reading some of the Meson documentation (Which was pretty helpful actually! They have [a very useful chart](https://mesonbuild.com/Builtin-options.html) that pointed to towards the parameter I needed to pass in order to get both the static and dynamic library.
> 
> As I am also a meson developer, thanks. :)

Oh well, what a coincidence :sweat_smile: ! To be honest, I have only used CMake and Makefiles; I haven't used Meson yet. However, from what I've seen, I really hope both of those projects took a note from Meson's documentation book. 
 
> > I don't know if I messed up, or it's simply glib's nature/shortcomings. Time to look towards `muslc`?
> 
> This is basically weird because glibc wants to dlopen NSS libraries in order to support the use case where you compile an executable on your laptop and then scp it to a server running LDAP where it needs to query LDAP in order to figure out what user accounts are available so it can retrieve networked records for `getpwuid_r(32791)`.
> 
> You could also have an alternative DNS resolver that may be needed in practice to actually _resolve_ stuff.
> 
> Anyway, there is some interesting conversation at https://developers.redhat.com/articles/2023/08/31/how-we-ensure-statically-linked-applications-stay-way
> 
> musl essentially "solves" this by not supporting those use cases, which may be a good thing in the event that it is mission-critical for you to avoid loading nss modules and it is acceptable to simply not support operation on systems where ignoring NSS leads to failure.
> 
> glibc defines a configure option:
> 
> > --enable-static-nss
> > > Compile static versions of the NSS (Name Service Switch) libraries. This is not recommended because it defeats the purpose of NSS; a program linked statically with the NSS libraries cannot be dynamically reconfigured to use a different name database.
> 
> and the associated FAQ: https://sourceware.org/glibc/wiki/FAQ#Even_statically_linked_programs_need_some_shared_libraries_which_is_not_acceptable_for_me.__What_can_I_do.3F

Thank you very much for those links! Very interesting reads. I was not familiar with NSS :D.


One last question: 

> The fact that net-misc/curl has this USE flag, depends on libpsl, and libpsl lacks the same USE flag, feels like a compelling argument. Thanks for the explanation!

Let's say that `libcurl` didn't depend on `libpsl` and that `libpsl` was "just another library". If the library supports it, shouldn't the USE flag be made available "always"? Or are there any drawbacks? Is it a "given the use flag, more bug reports will be made which takes time away from maintainers"-type issue? Or is it something else?

Asking just to get some insight.

- Best regards

**Comment by eli-schwartz** [comment]
> Let's say that `libcurl` didn't depend on `libpsl` and that `libpsl` was "just another library". If the library supports it, shouldn't the USE flag be made available "always"? Or are there any drawbacks? Is it a "given the use flag, more bug reports will be made which takes time away from maintainers"-type issue? Or is it something else?

Gentoo discourages use of static libraries in general: https://projects.gentoo.org/qa/policy-guide/installed-files.html#pg0302 It's not *usually* necessary for other gentoo packages, and adds complexity e.g. for propagating static-libs USE deps to other packages.

By default, every single software project distributing a library that uses either automake or meson "supports it" so we'd have to add the USE flag to most packages even though they typically are not actually needed. Sometimes they are needed, e.g. for building qemu and a few other things.

**Comment by lima-limon-inc** [comment]
> Gentoo discourages use of static libraries in general: https://projects.gentoo.org/qa/policy-guide/installed-files.html#pg0302 It's not _usually_ necessary for other gentoo packages, and adds complexity e.g. for propagating static-libs USE deps to other packages.
> 
> By default, every single software project distributing a library that uses either automake or meson "supports it" so we'd have to add the USE flag to most packages even though they typically are not actually needed. Sometimes they are needed, e.g. for building qemu and a few other things.

Great, thanks for the explanation!

PS: I don't know why github complains about a conflic. It's not showing me the conflict itself :/. Regardless, I'll update the commit if further modifications need to be made.

Best regards

**Comment by eli-schwartz** [comment]
> PS: I don't know why github complains about a conflic. It's not showing me the conflict itself :/.

It is because you did not `git rebase origin/master` and the parent commit for your change is from a month ago. On Sunday, the file you *moved* was modified via 352cd0de6fe03e87415f1cb55fbcca9c7295d98a.

> Regardless, I'll update the commit if further modifications need to be made.

What comes to mind is:
- the newly added USE flag implements new functionality, whereby having the USE flag *disabled* is equivalent to not having the USE flag
- the newly added USE flag defaults to disabled (no `+static-libs`, no profile `package.use`)

hence, for people who sync the repository and do not enable USE=static-libs, their installed package is semantically equivalent before and after the addition to IUSE. `emerge -uDU @world` will refrain from reinstalling libpsl, and it is right to refrain. This assumes that it is modified in-place rather than bumping to -r1.

In this scenario there is no value-add to bumping to -r1.

**Comment by lima-limon-inc** [comment]
> > PS: I don't know why github complains about a conflic. It's not showing me the conflict itself :/.
> 
> It is because you did not `git rebase origin/master` and the parent commit for your change is from a month ago. On Sunday, the file you _moved_ was modified via [352cd0d](https://github.com/gentoo/gentoo/commit/352cd0de6fe03e87415f1cb55fbcca9c7295d98a).
> 
> > Regardless, I'll update the commit if further modifications need to be made.
> 
> What comes to mind is:
> 
>     * the newly added USE flag implements new functionality, whereby having the USE flag _disabled_ is equivalent to not having the USE flag
> 
>     * the newly added USE flag defaults to disabled (no `+static-libs`, no profile `package.use`)
> 
> 
> hence, for people who sync the repository and do not enable USE=static-libs, their installed package is semantically equivalent before and after the addition to IUSE. `emerge -uDU @world` will refrain from reinstalling libpsl, and it is right to refrain. This assumes that it is modified in-place rather than bumping to -r1.
> 
> In this scenario there is no value-add to bumping to -r1.

Noted. I believe I addresed both issues on the latest commit. I removed the revision bump and synced the git history correctly (I think).

LMKWYT :blush: 

Best regards

**Comment by eli-schwartz** [comment]
```
>>> Installing (1 of 1) net-libs/libpsl-0.21.5::gentoo
 * CMP: =net-libs/libpsl-0.21.5 with net-libs/libpsl-0.21.5/image
 *  FILES:+usr/lib64/libpsl.a
 *   SIZE: 0.13MiB -> 0.27MiB, 12 -> 13 files
 * ------> FILES(+1) SIZE(+96.51%)

>>> Completed (1 of 1) net-libs/libpsl-0.21.5::gentoo
```

;)

**Comment by lima-limon-inc** [comment]
> ```
> >>> Installing (1 of 1) net-libs/libpsl-0.21.5::gentoo
>  * CMP: =net-libs/libpsl-0.21.5 with net-libs/libpsl-0.21.5/image
>  *  FILES:+usr/lib64/libpsl.a
>  *   SIZE: 0.13MiB -> 0.27MiB, 12 -> 13 files
>  * ------> FILES(+1) SIZE(+96.51%)
> 
> >>> Completed (1 of 1) net-libs/libpsl-0.21.5::gentoo
> ```
> 
> ;)

\:D/ 

If anything, it works on both of our machines haha :D.

**Comment by a17r** [comment]
> If the library supports it, shouldn't the USE flag be made available "always"? 

In general, no, that's not sufficient just by itself. What is requiring `static-libs` on this package?

**Comment by lima-limon-inc** [comment]
> In general, no, that's not sufficient just by itself to justify IUSE exposure. What is requiring `static-libs` on this package?

Hi @a17r . Here's the full answer: https://github.com/gentoo/gentoo/pull/38910#issuecomment-2398790472

Here's the TLDR:

> TLDR: I wanted to build a static executable that had libcurl statically. IIANM, libpsl was both needed and did not have a static-libs use flag.

**Comment by eli-schwartz** [comment]
> > If the library supports it, shouldn't the USE flag be made available "always"?
> 
> In general, no, that's not sufficient just by itself to justify IUSE exposure. What is requiring `static-libs` on this package?

I don't really understand the second half of your comment in the context of the complete comment you are replying to:

> Let's say that `libcurl` didn't depend on `libpsl` and that `libpsl` was "just another library". If the library supports it, shouldn't the USE flag be made available "always"? Or are there any drawbacks? Is it a "given the use flag, more bug reports will be made which takes time away from maintainers"-type issue? Or is it something else?

(And that, itself, is a direct reply to my saying the *first* half of your comment.)

**Comment by lima-limon-inc** [comment]
Great. Thank you @eli-schwartz for all the help and guidance! 

Have a great week :blush:
